src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -0,0 +1,412 @@
+
+      container.endpoints.foreach { endpoint =>
+        val portMapping = mesos.NetworkInfo.PortMapping.newBuilder
+
+        endpoint.containerPort.foreach(portMapping.setContainerPort)
+        endpoint.hostPort.foreach(portMapping.setHostPort)
+        endpoint.protocol.foreach(portMapping.setProtocol)
+
+        networkInfo.addPortMappings(portMapping)
+
+        // TODO: labels?
+      }
+    }
+
+    container.volumeMounts.foreach { volumeMount =>
+      podVolumes.find(_.name == volumeMount.name).map { hostVolume =>
+        val volume = mesos.Volume.newBuilder
+          .setContainerPath(volumeMount.mountPath)
+
+        hostVolume.host.foreach(volume.setHostPath)
+
+        // Read-write mode will be used when the "readOnly" option isn't set.
+        val mode = volumeMount.readOnly.getOrElse(false) match {
+          case true => mesos.Volume.Mode.RO
+          case false => mesos.Volume.Mode.RW
+        }
+
+        volume.setMode(mode)
+
+        containerInfo.addVolumes(volume)
+      }
+    }
+
+    container.image.foreach { im =>
+      val image = mesos.Image.newBuilder
+
+      im.forcePull.foreach(forcePull => image.setCached(!forcePull))
+
+      im.kind match {
+        case raml.ImageType.Docker =>
+          val docker = mesos.Image.Docker.newBuilder.setName(im.id)
+
+          image.setType(mesos.Image.Type.DOCKER).setDocker(docker)
+        case raml.ImageType.Appc =>
+          // These labels are necessary for AppC images to work.
+          // Given that Docker only works under linux with 64bit,
+          // let's (for now) set these values to reflect that.
+          val labels = mesos.Labels.newBuilder
+            .addAllLabels(
+              Iterable(
+              mesos.Label.newBuilder.setKey("os").setValue("linux").build,
+              mesos.Label.newBuilder.setKey("arch").setValue("amd64").build
+            ).asJava)
+
+          val appc = mesos.Image.Appc.newBuilder.setName(im.id).setLabels(labels)
+
+          image.setType(mesos.Image.Type.APPC).setAppc(appc)
+      }
+
+      val mesosInfo = mesos.ContainerInfo.MesosInfo.newBuilder.setImage(image)
+      containerInfo.setMesos(mesosInfo)
+    }
+
+    containerInfo
+  }
+
+  private[this] def computeHealthCheck(
+    healthCheck: raml.HealthCheck,
+    endpoints: Seq[raml.Endpoint]): mesos.HealthCheck.Builder = {
+    val builder = mesos.HealthCheck.newBuilder
+
+    if (healthCheck.command.isDefined) {
+      builder.setType(mesos.HealthCheck.Type.COMMAND)
+
+      val command = mesos.CommandInfo.newBuilder
+
+      healthCheck.command.get.command match {
+        case raml.ShellCommand(shell) =>
+          command.setShell(true)
+          command.setValue(shell)
+        case raml.ArgvCommand(argv) =>
+          command.setShell(false)
+          if (argv.nonEmpty) {
+            command.setValue(argv.head)
+          }
+
+          if (argv.size > 1) {
+            command.addAllArguments(argv.tail.asJava)
```

Comment:
are we sure that argv.tail is correct? posix normally wants arguments to include argv[0].. and this was carried over to mesos for normal task command execution. not sure if this semantic was maintained for health checks